### ISSUE TITLE
coc-clangd has magically stopped working

### ISSUE THREAD

**Issue by ricciolino** [clangd/coc-clangd#109]
Right yesterday all was working fine, today, with opening a cpp file on vim 8.2, autocmpletion and all the other stuff by coc-clangd are gone. I'm not very expert to investigate what is going on, I hope you could help me to solve my problem! Thanks in advance. I'm on Ubuntu 18.04.5 LTS and here is my settings:

$ uname -r
5.4.0-52-generic

$ clangd --version
clangd version 10.0.0-4ubuntu1~18.04.2

- CocInfo output:
2020-11-11T20:44:06.129 INFO (pid:668) [services] - registered service "clangd"
2020-11-11T20:44:06.132 INFO (pid:668) [services] - clangd state change: stopped => starting
2020-11-11T20:44:06.135 INFO (pid:668) [language-client-index] - Language server "clangd" started with 687
2020-11-11T20:44:06.138 ERROR (pid:668) [language-client-client] - connection error: Error: write EPIPE
    at afterWriteDispatched (internal/stream_base_commons.js:156:25)
    at writeGeneric (internal/stream_base_commons.js:147:3)
    at Socket._writeGeneric (net.js:788:11)
    at Socket._write (net.js:800:8)
    at doWrite (_stream_writable.js:403:12)
    at writeOrBuffer (_stream_writable.js:387:5)
    at Socket.Writable.write (_stream_writable.js:318:11)
    at StreamMessageWriter.write (/home/ricciolino/.vim/plugged/coc.nvim/build/index.js:18391:27)
    at /home/ricciolino/.vim/plugged/coc.nvim/build/index.js:17539:35
    at new Promise (<anonymous>)
    at Object.sendRequest (/home/ricciolino/.vim/plugged/coc.nvim/build/index.js:17529:26)
    at Object.initialize (/home/ricciolino/.vim/plugged/coc.nvim/build/index.js:74367:44)
    at LanguageClient.doInitialize (/home/ricciolino/.vim/plugged/coc.nvim/build/index.js:76483:27)
    at LanguageClient.initialize (/home/ricciolino/.vim/plugged/coc.nvim/build/index.js:76479:25)
    at /home/ricciolino/.vim/plugged/coc.nvim/build/index.js:76401:25 {
  errno: 'EPIPE',
  code: 'EPIPE',
  syscall: 'write'
} undefined
2020-11-11T20:44:06.140 INFO (pid:668) [services] - clangd state change: starting => stopped
2020-11-11T20:44:06.140 ERROR (pid:668) [language-client-client] - connection error: Error [ERR_STREAM_DESTROYED]: Cannot call write after a stream was destroyed
    at doWrite (_stream_writable.js:399:19)
    at writeOrBuffer (_stream_writable.js:387:5)
    at Socket.Writable.write (_stream_writable.js:318:11)
    at StreamMessageWriter.write (/home/ricciolino/.vim/plugged/coc.nvim/build/index.js:18393:27)
    at /home/ricciolino/.vim/plugged/coc.nvim/build/index.js:17539:35
    at new Promise (<anonymous>)
    at Object.sendRequest (/home/ricciolino/.vim/plugged/coc.nvim/build/index.js:17529:26)
    at Object.initialize (/home/ricciolino/.vim/plugged/coc.nvim/build/index.js:74367:44)
    at LanguageClient.doInitialize (/home/ricciolino/.vim/plugged/coc.nvim/build/index.js:76483:27)
    at LanguageClient.initialize (/home/ricciolino/.vim/plugged/coc.nvim/build/index.js:76479:25)
    at /home/ricciolino/.vim/plugged/coc.nvim/build/index.js:76401:25 {
  code: 'ERR_STREAM_DESTROYED'
} undefined
2020-11-11T20:44:06.141 ERROR (pid:668) [language-client-client] - connection error: Error [ERR_STREAM_DESTROYED]: Cannot call write after a stream was destroyed
    at doWrite (_stream_writable.js:399:19)
    at writeOrBuffer (_stream_writable.js:387:5)
    at Socket.Writable.write (_stream_writable.js:318:11)
    at StreamMessageWriter.write (/home/ricciolino/.vim/plugged/coc.nvim/build/index.js:18391:27)
    at /home/ricciolino/.vim/plugged/coc.nvim/build/index.js:17539:35
    at new Promise (<anonymous>)
    at Object.sendRequest (/home/ricciolino/.vim/plugged/coc.nvim/build/index.js:17529:26)
    at Object.shutdown (/home/ricciolino/.vim/plugged/coc.nvim/build/index.js:74368:36)
    at /home/ricciolino/.vim/plugged/coc.nvim/build/index.js:76575:31 {
  code: 'ERR_STREAM_DESTROYED'
} undefined
2020-11-11T20:44:06.141 ERROR (pid:668) [language-client-client] - connection error: Error [ERR_STREAM_DESTROYED]: Cannot call write after a stream was destroyed
    at doWrite (_stream_writable.js:399:19)
    at writeOrBuffer (_stream_writable.js:387:5)
    at Socket.Writable.write (_stream_writable.js:318:11)
    at StreamMessageWriter.write (/home/ricciolino/.vim/plugged/coc.nvim/build/index.js:18393:27)
    at /home/ricciolino/.vim/plugged/coc.nvim/build/index.js:17539:35
    at new Promise (<anonymous>)
    at Object.sendRequest (/home/ricciolino/.vim/plugged/coc.nvim/build/index.js:17529:26)
    at Object.shutdown (/home/ricciolino/.vim/plugged/coc.nvim/build/index.js:74368:36)
    at /home/ricciolino/.vim/plugged/coc.nvim/build/index.js:76575:31 {
  code: 'ERR_STREAM_DESTROYED'
} undefined
2020-11-11T20:44:07.096 WARN (pid:668) [extensions] - Extension coc-clangd activate cost more than 1s
2020-11-11T20:44:07.097 INFO (pid:668) [plugin] - coc.nvim 0.0.79-d3293597d1 initialized with node: v12.19.0 after 1038ms
2020-11-11T20:44:07.147 INFO (pid:668) [completion-complete] - Results from: around
2020-11-11T20:44:08.017 INFO (pid:668) [completion-complete] - Results from: around
2020-11-11T20:44:08.152 INFO (pid:668) [completion-complete] - Results from: around
2020-11-11T20:44:08.190 INFO (pid:668) [completion-complete] - Results from: around

- CocList -> extentions:
  * coc-clangd 0.4.14 ~/.config/coc/extensions/node_modules/coc-clangd                                    
  + coc-vimtex 1.0.4 ~/.config/coc/extensions/node_modules/coc-vimtex
  + coc-tsserver 1.5.5 ~/.config/coc/extensions/node_modules/coc-tsserver
  + coc-python 1.2.12 ~/.config/coc/extensions/node_modules/coc-python
  + coc-json 1.2.6 ~/.config/coc/extensions/node_modules/coc-json

- My CocConfig:
{
    "python.linting.pylintEnabled": true,
    "python.linting.enabled": true,
    "python.linting.pylintArgs": ["--extension-pkg-whitelist=cv2","--generated-members=torch.*"],
    "clangd.enabled": true,
    "clangd.path": "/home/ricciolino/.config/coc/extensions/coc-clangd-data/install/10.0.0/clangd_10.0.0/bin/clangd",
    "clangd.arguments": ["-std=c++17"]
}

**Comment by fannheyward** [comment]
Can't reproduce, try upgrade your coc.nvim to latest release.

**Comment by ricciolino** [comment]
I upgrade coc.nvim to the latest release, but I cannot get coc-clangd working. Any suggestion?

**Comment by ricciolino** [comment]
I try to reinstall coc.nvim (with Plug 'neoclide/coc.nvim', {'branch': 'release'} ), then I reinstalled coc-clangd (:CocInstall coc-clangd) but when I attempt to install the latest release of clangd (:CocCommand clangd.install) I get a "client coc abnormal exit with: -1"

**Comment by ricciolino** [comment]
I solved by removing clangd-10 from my system:

$ sudo apt remove --purge --auto-remove clangd-10

Then I install again coc-clangd (:CocInstall coc-clangd) and the latest release of clangd  (:CocCommand clangd.install).

**Comment by rcornall** [comment]
Getting the same stacktrace,

$ uname -r
4.15.0-123-generic
$ clangd --version
clangd version 11.0.0 (https://github.com/llvm/llvm-project 176249bd6732a8044d457092ed932768724a6f06)
$ node --version
v14.15.1

```
## versions

vim version: NVIM v0.5.0-828-g0a95549d6
node version: v14.15.1
coc.nvim version: 0.0.79-1e1f452a03
coc.nvim directory: /home/rcornall/.vim/plugged/coc.nvim
term: xterm-256color
platform: linux

## Log of coc.nvim

2020-11-17T12:36:42.952 INFO (pid:10610) [plugin] - coc.nvim 0.0.79-1e1f452a03 initialized with node: v14.15.1 after 89ms
2020-11-17T12:36:47.589 INFO (pid:10610) [services] - registered service "clangd"
2020-11-17T12:36:47.590 INFO (pid:10610) [services] - clangd state change: stopped => starting
2020-11-17T12:36:47.595 INFO (pid:10610) [language-client-index] - Language server "clangd" started with 10731
2020-11-17T12:36:47.615 INFO (pid:10610) [services] - clangd state change: starting => stopped
2020-11-17T12:36:47.615 INFO (pid:10610) [services] - clangd state change: stopped => starting
2020-11-17T12:36:47.620 INFO (pid:10610) [language-client-index] - Language server "clangd" started with 10732
2020-11-17T12:36:47.624 ERROR (pid:10610) [language-client-client] - connection error: Error: write EPIPE
    at afterWriteDispatched (internal/stream_base_commons.js:156:25)
    at writeGeneric (internal/stream_base_commons.js:147:3)
    at Socket._writeGeneric (net.js:785:11)
    at Socket._write (net.js:797:8)
    at writeOrBuffer (_stream_writable.js:352:12)
    at Socket.Writable.write (_stream_writable.js:303:10)
    at StreamMessageWriter.write (/home/rcornall/.vim/plugged/coc.nvim/build/index.js:18418:27)
    at /home/rcornall/.vim/plugged/coc.nvim/build/index.js:17566:35
    at new Promise (<anonymous>)
    at Object.sendRequest (/home/rcornall/.vim/plugged/coc.nvim/build/index.js:17556:26)
    at Object.initialize (/home/rcornall/.vim/plugged/coc.nvim/build/index.js:79591:44)
    at LanguageClient.doInitialize (/home/rcornall/.vim/plugged/coc.nvim/build/index.js:81706:27)
    at LanguageClient.initialize (/home/rcornall/.vim/plugged/coc.nvim/build/index.js:81702:25)
    at /home/rcornall/.vim/plugged/coc.nvim/build/index.js:81624:25 {
  errno: -32,
  code: 'EPIPE',
  syscall: 'write'
} undefined
2020-11-17T12:36:47.625 INFO (pid:10610) [services] - clangd state change: starting => stopped
2020-11-17T12:36:48.508 WARN (pid:10610) [extensions] - Extension coc-clangd activate cost more than 1s

vim version: NVIM v0.5.0-828-g0a95549d6
node version: v14.15.1
coc.nvim version: 0.0.79-1e1f452a03
coc.nvim directory: /home/rcornall/.vim/plugged/coc.nvim
term: xterm-256color
platform: linux

## Log of coc.nvim

2020-11-17T12:36:42.952 INFO (pid:10610) [plugin] - coc.nvim 0.0.79-1e1f452a03 initialized with node: v14.15.1 after 89ms
2020-11-17T12:36:47.589 INFO (pid:10610) [services] - registered service "clangd"
2020-11-17T12:36:47.590 INFO (pid:10610) [services] - clangd state change: stopped => starting
2020-11-17T12:36:47.595 INFO (pid:10610) [language-client-index] - Language server "clangd" started with 10731
2020-11-17T12:36:47.615 INFO (pid:10610) [services] - clangd state change: starting => stopped
2020-11-17T12:36:47.615 INFO (pid:10610) [services] - clangd state change: stopped => starting
2020-11-17T12:36:47.620 INFO (pid:10610) [language-client-index] - Language server "clangd" started with 10732
2020-11-17T12:36:47.624 ERROR (pid:10610) [language-client-client] - connection error: Error: write EPIPE
    at afterWriteDispatched (internal/stream_base_commons.js:156:25)
    at writeGeneric (internal/stream_base_commons.js:147:3)
    at Socket._writeGeneric (net.js:785:11)
    at Socket._write (net.js:797:8)
    at writeOrBuffer (_stream_writable.js:352:12)
    at Socket.Writable.write (_stream_writable.js:303:10)
    at StreamMessageWriter.write (/home/rcornall/.vim/plugged/coc.nvim/build/index.js:18418:27)
    at /home/rcornall/.vim/plugged/coc.nvim/build/index.js:17566:35
    at new Promise (<anonymous>)
    at Object.sendRequest (/home/rcornall/.vim/plugged/coc.nvim/build/index.js:17556:26)
    at Object.initialize (/home/rcornall/.vim/plugged/coc.nvim/build/index.js:79591:44)
    at LanguageClient.doInitialize (/home/rcornall/.vim/plugged/coc.nvim/build/index.js:81706:27)
    at LanguageClient.initialize (/home/rcornall/.vim/plugged/coc.nvim/build/index.js:81702:25)
    at /home/rcornall/.vim/plugged/coc.nvim/build/index.js:81624:25 {
  errno: -32,
  code: 'EPIPE',
  syscall: 'write'
} undefined
2020-11-17T12:36:47.625 INFO (pid:10610) [services] - clangd state change: starting => stopped
2020-11-17T12:36:48.508 WARN (pid:10610) [extensions] - Extension coc-clangd activate cost more than 1s
```

I noticed it's only failing to start in some directories, with larger projects. If I start a simple test.cpp somewhere it works OK, 
Eg:

```
## versions

vim version: NVIM v0.5.0-828-g0a95549d6
node version: v14.15.1
coc.nvim version: 0.0.79-1e1f452a03
coc.nvim directory: /home/rcornall/.vim/plugged/coc.nvim
term: xterm-256color
platform: linux

## Log of coc.nvim

2020-11-17T12:48:55.613 INFO (pid:14928) [plugin] - coc.nvim 0.0.79-1e1f452a03 initialized with node: v14.15.1 after 90ms
2020-11-17T12:49:03.709 INFO (pid:14928) [services] - registered service "clangd"
2020-11-17T12:49:03.710 INFO (pid:14928) [services] - clangd state change: stopped => starting
2020-11-17T12:49:03.715 INFO (pid:14928) [language-client-index] - Language server "clangd" started with 14973
2020-11-17T12:49:03.737 INFO (pid:14928) [services] - clangd state change: starting => running
2020-11-17T12:49:03.748 INFO (pid:14928) [services] - service clangd started
```

I have an empty CocConfig.

**Comment by ricciolino** [comment]
Have you tried to
• Uninstall the coc-clangd extension:

> `:CocUninstall coc-clangd`

• Remove the "coc-clangd-data" folder inside the `$HOME/.config/coc/.....` path (can't remember the precise folder name and the precise path right now, if you browse a bit you'll find a folder with your current clangd installation)
• Uninstall and Reinstall the coc-nvim plugin thought VimPlug (Comment and Uncomment inside the `.vimrc` file)
• Reinstall the `coc-clangd` extension

> `:CocInstall coc-clangd`

• Reinstall the latest clangd version:

> `:CocCommand clangd.install`

In my case this solve the issue !

**Comment by rcornall** [comment]
Actually I did those steps almost exactly, including forcing re-install of coc-nvim (with :PlugClean + :PlugInstall).
Clangd I installed manually.

I downgraded to coc.nvim  v0.0.78, and have found root cause, I get some nicer errors:

```
## Output channel: clangd

clangd: Unknown command line argument '-–query-driver'.  Try: '/home/rcornall/.config/coc/extensions/coc-clangd-data/install/11.0.0/clangd_11.0.0/bin/clangd --help'
clangd: Did you mean '--query-driver'?
clangd: Unknown command line argument '/home/rcornall/wd/****hiding..this..****sysroots/x86_64/....'.  Try: '/home/rcornall/.config/coc/extensions/coc-clangd-data/install/11.0.0/clangd_11.0.0/bin/clangd --help'
```

So,

1. 1st issue was I can see It's reading my  `../.vim/coc-settings.json` from some upper parent directory which I didn't realize it would do.
2. 2nd issue is I somehow had some weird hyphen character '-–' in those settings which caused clangd to fail.

But 3rd issue is why don't these nice clangd logs show up on the later coc.nvim release though.. maybe a coc.nvim issue?

**Comment by ricciolino** [comment]
Well my knowledge is out of scope in this case...it would be nice if someone could explain what cause this strange behavior!

**Comment by fannheyward** [comment]
@rcornall It's not recommend to downgrade coc.nvim, you'd better to use latest release.

> But 3rd issue is why don't these nice clangd logs

You can get clangd logs in `:CocCommand workspace.showOutput`. There's something broken in coc-clangd 0.5, I've just released v0.5.2 to fix this.

**Comment by wangsen992** [comment]
I don't think this follows the line along the current solution in the thread. Since I'm installing everything on HPC (no sudo rights), I installed my own _clangd_ (v12.0.1) on my home dir, then inspired by @ricciolino I found `coc-clangd-data` folder, then replaced the `coc-clangd-data/install/12.0.0/clangd_12.0.0/bin/clangd` with a symlink of my own clangd install, then i got it working
